tls/s2n_connection.c

```diff
@@ -980,6 +980,21 @@ const char *s2n_connection_get_cipher(struct s2n_connection *conn)
     return conn->secure.cipher_suite->name;
 }
 
+int s2n_connection_get_cipher_iana_value(struct s2n_connection *conn, uint8_t *first, uint8_t *second)
+{
+    ENSURE_POSIX_REF(conn);
+    /* ensure we've negotiated a cipher suite */
+    ENSURE_POSIX(conn->secure.cipher_suite != NULL, S2N_ERR_INVALID_STATE);
```

Comment:
S2N_ERR_INVALID_STATE is classified as an S2N_ERR_T_INTERNAL. Should it be a usage error instead? Here it's more about how the API is being called.

Reply:
hmm... The description doesn't describe it as internal:

https://github.com/awslabs/s2n/blob/c8d04e4611bb44e3a09bcf991fec1c6ee4282366/error/s2n_errno.c#L245

Is there another error type that I should use? Should I define a new one?

Reply:
Or should we reclassify `S2N_ERR_INVALID_STATE` as a usage error?

Reply:
It looks like it's currently used by the certificate validation logic, where I would say it is an internal error (the user isn't really aware of the internal validation state, and it's triggered by the handshake, not the user). Like, if the user gets S2N_ERR_INVALID_STATE, that would be a sign there's a bug in the S2N code.

There should probably be two separate errors, rather than just making S2N_ERR_INVALID_STATE a usage error.

Reply:
Isn't S2N_ERR_NULL better suited for this use-case? 
We could use `ENSURE_REF(conn->secure.cipher_suite)` that raises S2N_ERR_NULL error code when NULL value is encountered.

Reply:
> It looks like it's currently used by the certificate validation logic, where I would say it is an internal error

The error message is terribly worded if that's the case. It specifically calls out invalid use of an API and that the caller should read the API documentation as to why it happened. How is that internal?

> Isn't S2N_ERR_NULL better suited for this use-case?

That wouldn't tell the user anything. This is specifically in place to let them know that they tried to get the negotiated cipher before the connection had actually negotiated it. I think both the error name (`S2N_ERR_INVALID_STATE`) and description reflect this use case perfectly.

Reply:
I've removed the usage of that error code in the validator and reclassified it as usage